Added the ability to build from a Dockerfile

ClaytonNorthey92 opened the pull request:
Fixes #57

Added the ability to build from a Dockerfile, used an embedded struct called `FromDockerfile` as part of the `ContainerRequest` struct.

If you specify a `Context`, it will use that to build the image.  It will not allow you to specify both a `Context` and an `Image`

Added to README with an example.

ClaytonNorthey92 commented:
hey @gianarb it looks like the build is timing out here, when I run the tests locally they finish in ~50 seconds, even without caching.

Is there a good way to debug the travis build?  

```
Claytons-MacBook-Pro-2:testcontainers-go claytonnorthey$ git log -1
commit 9233e2cd492fc3fb5e8370fc2c00c31b4458733c (HEAD -> feature/from-dockerfile, origin/feature/from-dockerfile)
Author: ClaytonNorthey92 <cnorthey@interactions.com>
Date:   Thu Aug 8 21:42:29 2019 -0400

    Added the ability to build from a Dockerfile
    Fixes #57
Claytons-MacBook-Pro-2:testcontainers-go claytonnorthey$ go test ./...
ok  	github.com/testcontainers/testcontainers-go	49.657s
?   	github.com/testcontainers/testcontainers-go/wait	[no test files]
Claytons-MacBook-Pro-2:testcontainers-go claytonnorthey$ go test -count=1 ./...
ok  	github.com/testcontainers/testcontainers-go	50.338s
?   	github.com/testcontainers/testcontainers-go/wait	[no test files]
Claytons-MacBook-Pro-2:testcontainers-go claytonnorthey$ go version
go version go1.11.12 darwin/amd64
```

mdelapenya commented on container.go:
```diff
@@ -53,6 +67,7 @@ type ContainerRequest struct {
 	Privileged   bool   // for starting privileged container
 
 	SkipReaper bool // indicates whether we skip setting up a reaper for this
+	err        error
```
Is there a case to store the error in the struct? I would see it more convenient to return the error from the Validate method.

Wdyt?

ClaytonNorthey92 replied:
hmm I guess I don't have a strong one...I could change the `Validate` method to be something like this (pseudo code):
```go
var err error
for _, validateMethod := range []func() error {c.validateContextAndImage} {
    err = validateMethod()
    if err != nil {
        return err
    }
} 
return nil
```

then you could make the `validateContextAndImage` return an error, rather than modify the one on the struct.

Ultimately I wanted to have a clean way to validate the config, while making it easier to add more validation methods

mdelapenya replied:
I like this second approach, as we do not alter the structure with the validation status, but as you said, I do not have a strong opinion on that. And not being a maintainer 😄 !!

For me it would be a 👍

ClaytonNorthey92 replied:
okay thanks for the feedback , I like the second approach better too, glad you brought this up! I can update the pull request at some time today!

ClaytonNorthey92 replied:
okay, I have updated the pull request with what we talked about, thanks again!

ClaytonNorthey92 commented on README.md:
````diff
@@ -107,3 +107,19 @@ func TestRedisPing(t *testing.T) {
+
+```
+req := ContainerRequest{
+		FromDockerfile: FromDockerfile{
````
this might need to be `testcontainers.FromDockerfile` in the README...

mdelapenya commented:
Hi @gianarb @bsideup Is there anything we can do with this PR?

Thanks!

ClaytonNorthey92 commented:
just for some extra information:

I added a timeout to the tests on the build server of 2 minutes, the test I added is taking longer than that (over 10 minutes I think) so Travis is terminating it

I have tested this locally and my test finishes under 1 minute and terminates the containers properly

it is hard for me to debug what is happening on the build server, so I could use help with that

mdelapenya commented:
BTW @ClaytonNorthey92, did you take a look at https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received?

On the other hand it would make sense to detect why the timeout on Travis

ClaytonNorthey92 commented:
Good idea @mdelapenya ...I wonder if I can add some logging to the tests to see if either: they do get stuck, or there is just no output

ClaytonNorthey92 commented:
Ah looks like there might be an error pulling the redis image, I will use a different one, maybe one that is already used in the tests 🤔

ClaytonNorthey92 commented:
okay I got it passing...what fixed it was making sure I read the docker response into a buffer...I wonder if not doing so doesn't wait for the request to finish?

ClaytonNorthey92 commented on README.md:
```diff
@@ -61,3 +61,65 @@ can use it to make a GET: `resp, err := http.Get(fmt.Sprintf("http://%s", ip))`
 nginxC.Terminate(ctx, t)`. `t` is `*testing.T` and it is used to notify is the
 `defer` failed marking the test as failed.
+
+You can build more complex flow using env var to configure the containers. Let's
```
accidentally added this from rebasing incorrectly, will remove

gianarb commented:
I love this PR @ClaytonNorthey92 ! Thanks for your work and thans @mdelapenya for your support!
I will have a look at it today and I will give it a try! But after a quick look I can say it looks great

gianarb commented:
Hello! can you rebase this asl weel @ClaytonNorthey92 !? :) thanks a lot

ClaytonNorthey92 commented:
yup! just rebased @gianarb

gianarb commented:
Thanks a lot
